cli: refactors; make `cockroach-sql`'s CLI parameter handling match `cockroach sql`

knz opened the pull request:
Note to the reviewers:
- The PR has been split into many small commits to ease the review.
- The "main commit" is the last in the sequence; all commits before the last constitute a long-awaited refactor of the connection parameter handling, including moving much of the security-related logic away from `cli` into a sub-package of `security` where it belongs.
- We only have testing code to exercise the command-line handling for `cockroach sql`, which at this point should be considered sufficient given we're using the same config code in `cockroach-sql` now. To make tests work across the two commands, we would need to make progress on #80921 first.

Fixes #81882.
Fixes #82024.
Unblocks progress on #29285.

knz committed:
security: move file name strings to a new package `certnames`

We want to have the cert/key filenames in a package with minimal
dependencies, suitable for embedding in the lightweight SQL
executable. This commit starts the work by moving the file name
constants in a new separate package.

Release note: None

knz committed:
security: move the CertsLocator to the new package 'certnames'

Release note: None

knz committed:
security,rpc: avoid depending on CertManager in LoadSecurityOptions

This is needed to move the LoadSecurityOptions code to a lightweight
package independent of 'security'.

Release note: None

knz committed:
rpc: finish connecting LoadSecurityOptions to securityassets

Release note: None

knz committed:
rpc: detach LoadSecurityOptions and PGURL from SecurityContext

This paves the way to moving them to a standalone package.

Release note: None

knz committed:
cli,rpc: detach LoadSecurityOptions and PGURL from `base`

This paves the way further to move them to a separate package.

Release note: None

knz committed:
rpc: move LoadSecurityOptions and PGURL to a new lightweight package

This allows clarifies their scope as separate from the RPC logic.

Release note: None

knz committed:
cli/client_url: lift flag override in helper function

This is one step towards lifting client_url to a different package.

Release note: None

knz committed:
cli/client_url: lift flag override into helper function (2/)

This is one step towards lifting client_url to a different package.

Release note: None

knz committed:
cli/client_url: lift flag override into helper function (3/)

This is one step towards lifting client_url to a different package.

Release note: None

knz committed:
cli/client_url: lift flag override into helper function (4/)

This is one step towards lifting client_url to a different package.

Release note: None

knz committed:
cli/client_url: lift flag override into helper function (5/)

This is one step towards lifting client_url to a different package.

Release note: None

knz committed:
cli/client_url: lift the config logic into a standlone fn

This is one step towards lifting client_url to a different package.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82020)
<!-- Reviewable:end -->

knz committed:
clientconnurl: rename ClientOptions to ClientSecurityOptions

to make room for a non-security ClientOptions.

Release note: None

knz committed:
cli/client_url: move makeClientConnURL into a standalone fn

This is one more step towards lifting the logic to a lightweight
package.

Release note: None

knz committed:
cli: move makeClientConnURL to a standalone package

This also simplifies as follows:
- LoadSecurityOptions can now be un-exported.
- PGURL() can be simplified to use MakeClientConnURL() directly.

Release note: None

knz committed:
cli: move addrSetter into package `addr`

This makes it available for use by other go programs that
don't want to depend on `cli`, such as `cockroach-sql`.

Release note: None

knz committed:
cli: use clientconnurl.ClientOptions where appropriate

This simplifies the code and makes it more readable.

Release note: None

knz committed:
cli: move the URL parsing code to a standalone package

This is to prepare its use in the cockroach-sql standalone program.

Release note: None

knz committed:
cockroach-sql: increase compatibility with `cockroach-sql`

Release note (cli change): The standalone `cockroach-sql` executable
now has more compatibility with `cockroach sql`, so it can be used as
a drop-in replacement. For example, it supports running without a URL,
using connection defaults. It also supports overriding `--certs-dir`
and other client-side options also supported by `cockroach sql`.

knz commented:
NB: Reviewable seems to have issues displaying the entire commit sequence. The Github review interface is fine with it.

If the reviewer(s) would prefer that, I can split this PR into a sequence of small(er) PRs.

otan commented on pkg/cli/client_url.go:
```diff
@@ -160,6 +155,74 @@ func (u urlParser) setInternal(v string, warn bool) error {
+func updateURL(
+	newURL string,
+	sslStrict bool,
+	makeStrictErr func() error,
```
optional: how do you feel about these functions being an interface instead? 
could add unit tests too, if we wanted to. having to pass a lot of hooks in is easy to lose track of.

knz replied:
good idea. done (in split PR)

otan commented on pkg/cmd/cockroach-sql/main.go:
```diff
@@ -102,14 +152,45 @@ var sqlCmd = &cobra.Command{
+		return err
+	}
+
+	// How we're going to authenticate.
```
nit: comment sounds like a question - maybe "Get how we're going to authenticate"

otan reviewed (commented):
mostly some naming nits here, curious about how you feel about them

the `clientconnurl` package and `clienturl` package is a little similarly named and easy to get confused by. what are your thoughts on:

* clientconnurl -> pgconnurl (is it client specific?)
* clienturl -> cliurl, to make it clear only the CLI should be using it?

knz commented:
Yes I agree some names here are unfortunate.

I will split this PR so we can iterate on the review more easily.

knz commented:
I've split the PR into the following pieces:
- #82068
- #82069 
- #82070 
- #82071 
- #82072 
- #82074

knz commented on pkg/security/securityassets/security_assets.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+// AssetLoader describes the functions necessary to read certificate and key files.
+type AssetLoader struct {
```
done

knz commented on pkg/security/certnames/certs_locator.go:
```diff
@@ -0,0 +1,203 @@
+)
+
+// A CertsLocator provides locations to certificates.
+type CertsLocator struct {
```
agreed. Done (in split PR)

knz committed:
clientsecopts: rename PGURL to MakeURLForServer

We have enough things called "PGURL" as it is. This rename clarifies
the purpose.

Release note: None

knz committed:
certnames: rename IsCertificateFile to IsCertificateFilename

... since the function does not actually look at the contents of the
file.

Release note: None

knz committed:
clientsecopts: use an interface instead of callbacks for AnalyzeClientURL()

This documents the API semantics makes the code slightly easier to
reason about.

Release note: None

knz committed:
cliflagcfg: new package containing common flag def logic

This will be needed to ensure `cockroach-sql` supports the same env
vars.

Release note: None

knz committed:
cli/flag: move the flag definition logic to cliflagcfg

This ensures that the basic primitives are reusable out of package
`cli`.

Release note: None

knz committed:
cli: use cliflagcfg.FlagSetForCmd directly

Release note: None

knz committed:
cliflags: new package with the common client flag definitions

Release note: None

knz committed:
cliflagcfg: new variants of flag fns with depth arg

So that the caller can control at which level the env var
consumer check occurs.

(This can be explored via `cockroach debug env`.)

Release note: None

knz commented:
unrelated flake

bors r=otan,catj-cockroach

knz committed:
workloadccl: skip flaky TestImportFixtures

Release note: None

knz commented:
added a skip

bors r=otan,catj-cockroach
